Hw 5

yadarochka opened the pull request.

yadarochka committed:
изменена сборка проекта

yadarochka committed:
рефакторинг кода

yadarochka committed:
добавлен функционал

добавлен бесконечный скролл

yadarochka committed:
добавлен функционал

добавлен бесконечный скролл и выбор по категориям(при самом изменении дропдауна почему-то не происходит новый запрос)

yadarochka committed:
lint

yadarochka commented:
я сделал дропдаун со списком категорий по подобию дропдауна с валютой, но почему-то после выбора валюты происходит новый запрос, а после выбора категории нет =( не могу понять, что не так

yadarochka committed:
рефакторинг кода

yadarochka committed:
оптимизация + исправление багов

yadarochka committed:
изменения интерфейса

изменены страницы CoinListPage и CoinPage
добавлены новые UI-компоненты
доработаны старые UI-компонентов

andrewgul commented on README.md:
```diff
@@ -1,46 +1,13 @@
 
 Runs the app in the development mode.\
-Open [http://localhost:3000](http://localhost:3000) to view it in the browser.
+Open [http://127.0.0.1:9000](http://localhost:3000) to view it in the browser.
```
не поменял вторую ссылку
```suggestion
Open [http://127.0.0.1:9000](http://127.0.0.1:9000) to view it in the browser.
```

andrewgul commented on src/components/UI/Card/Card.tsx:
```diff
@@ -45,14 +46,12 @@ export const Card: FC<CardProps> = ({
   currency,
   coinData = [],
   coinLabels = [],
   onMouseEvent,
```
этот проп не используется

andrewgul commented on src/components/UI/Card/Card.tsx:
```diff
@@ -65,7 +64,7 @@ export const Card: FC<CardProps> = ({
         <h3 className={styles["card__title"]}>{name}</h3>
         <span className={styles["card__subtitle"]}>
-          {userType && `00.00 `}
+          {cardType === "userType" && `00.00 `}
```
зачем это нужно? компонент карточки используется со значением пропса `cardType="userType"` на экране просмотра конкретной валюты, при этом все значения в этой карточке по нулям.. может его не надо вообще на этом экране использовать?
<img width="439" alt="Снимок экрана 2022-09-14 в 13 03 03" src="https://user-images.githubusercontent.com/56839604/190125791-7c6f6bea-9175-45b4-a943-5ba416d0c9ab.png">

yadarochka replied:
я так понял в макете это для отображения количества валюты у конкретного пользователя, когда он в аккаунт заходит. что-то типа у вас столько-то биткоинов...

andrewgul replied:
а, понял тебя, вопрос исчерпан

andrewgul commented on src/components/UI/Card/Card.tsx:
```diff
@@ -79,44 +78,24 @@ export const Card: FC<CardProps> = ({
-            </div>
-          )}
+          <IncreaseOrDecrease
+            number={"00.00"}
```
```suggestion
            number="00.00"
```
но опять же (про что выше писал) не совсем понимаю, что это за нули

andrewgul commented on src/components/UI/Dropdown/Dropdown.tsx:
```diff
@@ -10,10 +13,13 @@ export const Dropdown: FC<MultiDropdownProps> = ({
 
-  const itemOnClick = (el: Options) => {
+  const toggleList = () => {
+    setIsVisible(!isVisible);
```
в сеттер состояния можно передавать не значение, а коллбэк, где аргумент – предыдущее значение состояния, а возвращаемое значение – новое состояние
```suggestion
    setIsVisible((prevIsVisible) => !prevIsVisible);
```
это полезно, когда следующее состояние зависит от предыдущего (как в данном случае)

andrewgul commented on src/components/UI/Dropdown/Dropdown.tsx:
```diff
@@ -22,25 +28,29 @@ export const Dropdown: FC<MultiDropdownProps> = ({
-        className={styles["dropdown__item"]}
+        className={classNames(
+          styles["dropdown__item"],
+          el.key === value.key ? styles["active"] : ""
```
```suggestion
          el.key === value.key && styles.active
```
можно так

andrewgul commented on src/components/UI/IncreaseOrDecrease/IncreaseOrDecrease.tsx:
```diff
@@ -0,0 +1,52 @@
+};
+
+export const IncreaseOrDecrease: FC<IncreaseOrDecreaseProps> = ({
+  procent,
```
заранее извиняюсь за такой душный коммент, но я думаю тут должно быть `percent`...

andrewgul commented on src/components/UI/IncreaseOrDecrease/IncreaseOrDecrease.tsx:
```diff
@@ -0,0 +1,52 @@
+  if (procent) {
+    number = procent;
+  }
+  if (typeof number === "string") {
+    return (
+      <div
+        className={classNames(
+          styles["inline-block"],
+          Number(number) === 0
+            ? styles.gray
+            : Number(number) > 0
+            ? styles.green
+            : styles.red,
+          className
+        )}
+      >
+        {number}
+        {procent ? "%" : ""}
+      </div>
+    );
+  } else if (number !== undefined) {
+    return (
+      <div
+        className={classNames(
+          styles["inline-block"],
+          number === 0 ? styles.gray : number > 0 ? styles.green : styles.red,
+          className
+        )}
+      >
+        {number}
+        {procent ? "%" : ""}
+      </div>
+    );
+  } else return <></>;
```
я вроде писал уже про это...
во-первых, зачем тут `else`-ы? можно просто
```
if (a) {
  return valueA;
}

if (b) {
  return valueB;
}

return valueC;
```
во-вторых, не надо возвращать пустой фрагмент, возвращай `null` в таких случаях

andrewgul replied:
это можно записать намного короче. отличие первых двух блоков - приводишь ли ты `number` к `Number`, все остальное у тебя одинаковое
```suggestion
if (!!number) {
    return (
      <div
        className={classNames(
          styles["inline-block"],
          Number(number) === 0 && styles.gray,
          Number(number) > 0 ? styles.green : styles.red,
          className
        )}
      >
        {number}
        {procent ? "%" : ""}
      </div>
    );
  }
  
  return null;
```

andrewgul commented on src/components/UI/IncreaseOrDecrease/IncreaseOrDecrease.tsx:
```diff
@@ -0,0 +1,52 @@
+  procent,
+  number,
+  className,
+}): JSX.Element => {
```
можно не прописывать тип возвращаемого значения, если ты указал тип функции... к тому же из-за того, что ты прописал тут `JSX.Element`, TS не дает вернуть `null`, хотя из компонента можно возвращать `null`

andrewgul commented on src/pages/CoinListPage/CoinListPage.tsx:
```diff
@@ -1,98 +1,157 @@
+          )}
+        >
+          Market is{" "}
+          {store.globalDataStore.globalData.marketCapChangePercentage24hUsd > 0
```
для этого можно сделать геттер в сторе

andrewgul commented on src/pages/CoinPage/CoinPage.tsx:
```diff
@@ -1,102 +1,121 @@
-                  ({store.coin.symbol.toUpperCase()})
+  if (store.meta === Meta.error) {
+    return <NotFoundPage />;
+  } else
```
про `else` уже писал

andrewgul commented on src/pages/NotFoundPage/NotFoundPage.tsx:
```diff
@@ -0,0 +1,22 @@
+export const NotFoundPage = () => {
+  const navigate = useNavigate();
+
+  const hundlerButtonClick = () => navigate("/");
```
думаю тут должно быть `handlerButtonClick`...

andrewgul commented on src/store/coinListStore/CategoryStore.ts:
```diff
@@ -0,0 +1,72 @@
+import { requestCategoryList } from "./requestCoinList";
+
+export default class CategoryStore implements ILocalStore {
+  _meta: Meta = Meta.initial;
+  options: CategoryModel[] = [];
+  _value: CategoryModel = {
```
```suggestion
  private _meta: Meta = Meta.initial;
  options: CategoryModel[] = [];
  private _value: CategoryModel = {
```
сделай их приватными раз уж назвал с нижнего подчеркивания и сделал для них геттеры

andrewgul commented on src/store/models/coinList/category.ts:
```diff
@@ -0,0 +1,36 @@
+export const normalizeCategoryModel = (
+  raw: CategoryApiModel[]
+): CategoryModel[] => {
+  const afdasfd: CategoryModel[] = [
```
переименуй пожалуйста))

andrewgul commented on webpack.config.js:
```diff
@@ -0,0 +1,109 @@
+        use: "babel-loader",
+      },
+      {
+        test: /\.(woff)$/,
```
можно сразу для большего кол-ва форматов шрифтов регулярку прописать
```suggestion
        test: /\.(eot|woff2|woff?)$/,
```

andrewgul commented on webpack.config.js:
```diff
@@ -0,0 +1,109 @@
+      fonts: path.join(srcPath, "store"),
+    },
+  },
+  devServer: {
+    host: "127.0.0.1",
+    port: "9000",
+    historyApiFallback:  true,
+  },
```
нет hot module replacement
```suggestion
  devServer: {
    host: "127.0.0.1",
    port: "9000",
    hot: true,
    historyApiFallback:  true,
  },
```

yadarochka commented:
> я сделал дропдаун со списком категорий по подобию дропдауна с валютой, но почему-то после выбора валюты происходит новый запрос, а после выбора категории нет =( не могу понять, что не так

а всё, разобрался, я забыл зависимости в useAsync передать....

andrewgul commented:
еще пара моментов:
- значения выбранных фильтров строит сохранять в query параметрах, чтобы можно было перейти по ссылке, и открылась страница с уже примененными фильтрами
- было бы неплохо добавить поиск по enter
- и реализовать сброс фильтров
- можешь попробовать разложить твой проект, чтобы на защите показывать его не локально, а по ссылке. можно попробовать сделать деплой на [netlify](https://www.netlify.com/blog/2016/09/29/a-step-by-step-guide-deploying-on-netlify/), например
![photo_2022-09-14_17-14-19](https://user-images.githubusercontent.com/56839604/190181045-b0188896-00ef-4bc5-9de5-3fab4858a8aa.jpg)

yadarochka committed:
добавлен функционал

можно выбрать интервал изменения цены валюты на её странице

yadarochka committed:
Update README.md

Co-authored-by: Andrew G. <56839604+andrewgul@users.noreply.github.com>

yadarochka committed:
новый функционал

теперь отображается удобный таймлайн на графике
добавлен хедер для всех страниц
добавлено бургер меню для навигации

yadarochka committed:
lint

yadarochka committed:
Merge branch 'hw-5' of https://github.com/yadarochka/kts-crypto-project into hw-5
